pkg/knowledge_base2/graph.go

```diff
@@ -235,16 +236,19 @@ func Upstream(dag construct.Graph, kb TemplateKB, rid construct.ResourceId, laye
 		return nil, fmt.Errorf("unknown layer %s", layer)
 	}
 	err := graph_addons.WalkUp(dag, rid, f)
-	return result, err
+	return result.ToSlice(), err
 }
 
 func layerWalkFunc(
 	dag construct.Graph,
 	kb TemplateKB,
 	rid construct.ResourceId,
 	layer DependencyLayer,
-	result []construct.ResourceId,
+	result set.Set[construct.ResourceId],
 ) (graph_addons.WalkGraphFunc[construct.ResourceId], error) {
+	if result == nil {
```

Comment:
If `result == nil` then this function doesn't really do anything since it doesn't return the set. Actually, is this function used? I'm not sure the `[]ResourceId` argument would work either.

Reply:
its used in here so that arguement wasnt actually needed for the reason it was being used. I think thats the only place its used but dont want to refactor that as a part of this pr if we think that needs to change
```
// LayeredUpstream returns the first resource that matches `selector` which is upstream of `resource` for the specified layer
func (ctx DynamicValueContext) LayeredDownstream(
	selector any,
	resource construct.ResourceId,
	layer string,
) (construct.ResourceId, error) {
	selId, err := TemplateArgToRID(selector)
	if err != nil {
		return construct.ResourceId{}, err
	}

	dependencyLayer := DependencyLayer(layer)
	f, err := layerWalkFunc(ctx.Graph, ctx.KnowledgeBase, resource, dependencyLayer, nil)
	if err != nil {
		return construct.ResourceId{}, err
	}
	result := construct.ResourceId{}
	wrapper := func(path graph_addons.Path[construct.ResourceId], nerr error) error {
		id := path[len(path)-1]
		if selId.Matches(id) {
			result = id
			return graph_addons.StopWalk
		}
		return f(path, nerr)
	}
	err = graph_addons.WalkDown(ctx.Graph, resource, wrapper)
	if err != nil {
		return construct.ResourceId{}, err
	}
	return result, nil
}

// Downstream returns the fi
```

Reply:
That's fine then. When we add tests for `LayeredDownstream` it should uncover any issues if there are any.